File: backend/app/schemas/slot.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field, model_validator


class SlotCreate(BaseModel):
    started_at: datetime
    ended_at: datetime
    raw_text: str = Field(..., max_length=100_000)
    metadata: dict[str, str | int | float | bool] | None = None

    @model_validator(mode="after")
    def check_dates(self) -> SlotCreate:
        if self.started_at and self.ended_at:
            if self.ended_at <= self.started_at:
                raise ValueError("ended_at must be after started_at")
        return self


class SlotResponse(BaseModel):
    id: str
    terminal_id: str | None
    tenant_id: str
    started_at: datetime
    ended_at: datetime
    duration_secs: int | None
    language: str | None
    word_count: int | None
    status: str
    tags: list[str] = []
    metadata: dict
    created_at: datetime
    score_overall: float | None = None

    model_config = {"from_attributes": True}


class SlotListResponse(BaseModel):
    items: list[SlotResponse]
    total: int
    page: int
    per_page: int


class SlotAccepted(BaseModel):
    slot_id: str
    status: str = "accepted"
    config: dict


class SlotDetail(SlotResponse):
    raw_text: str
    evaluation: "EvaluationResponse | None" = None


class ReEvaluateResponse(BaseModel):
    slot_id: str
    status: str = "re-evaluating"


class BulkReEvaluateRequest(BaseModel):
    slot_ids: list[str]


class BulkReEvaluateResponse(BaseModel):
    queued: int
    slot_ids: list[str]

File: portal/src/types/slot.ts
import { SlotStatus } from "./index";

export interface Slot {
  id: string;
  terminal_id: string;
  tenant_id: string;
  started_at: string;
  ended_at: string;
  duration_secs: number;
  language: string | null;
  word_count: number | null;
  status: SlotStatus;
  tags: string[];
  metadata: Record<string, unknown> | null;
  created_at: string;
  score_overall: number | null;
}

export interface SlotDetail extends Slot {
  raw_text: string;
  evaluation: Evaluation | null;
}

export interface Evaluation {
  id: string;
  slot_id: string;
  tenant_id: string;
  ai_provider: string;
  ai_model: string;
  prompt_version: string;
  score_overall: number | null;
  score_sentiment: number | null;
  score_politeness: number | null;
  score_compliance: number | null;
  score_resolution: number | null;
  score_upselling: number | null;
  score_response_time: number | null;
  score_honesty: number | null;
  sentiment_label: string | null;
  language_detected: string | null;
  summary: string | null;
  strengths: string[] | null;
  weaknesses: string[] | null;
  recommendations: string[] | null;
  unclear_items: string[] | null;
  flags: string[] | null;
  unavailable_items: string[] | null;
  swearing_count: number | null;
  swearing_instances: string[] | null;
  off_topic_count: number | null;
  off_topic_segments: string[] | null;
  speaker_segments: { speaker: string; text: string }[] | null;
  tokens_used: number | null;
  evaluation_duration_ms: number | null;
  is_unclear: boolean;
  created_at: string;
}

export interface SlotListParams {
  page?: number;
  per_page?: number;
  terminal_id?: string;
  status?: SlotStatus;
  date_from?: string;
  date_to?: string;
  min_score?: number;
  max_score?: number;
}

export interface BulkReEvaluateRequest {
  slot_ids: string[];
}
